ports/stm32/mboot/main.c

```diff
@@ -731,9 +952,227 @@ int do_write(uint32_t addr, const uint8_t *src8, size_t len) {
+            return -1; 
+        } else {
+            // Don't support flashing in un-encrypted mode
+            dfu_state.status = DFU_STATUS_ERROR_VENDOR;
+            dfu_state.error = MBOOT_ERROR_STR_UNENCRYPTED_IDX;
+            return -1;
+        }
+
+    } else {  // encryption_enabled()
+        if (addr == (uint32_t)APPLICATION_ADDR) {
+            // Base address of main firmware, reset any previous state
+            fw_enc_buff_addr = fw_enc_buff_next = 0;
+            next_page_erase = addr;
+        }
+
+        int32_t buff_addr = fw_enc_buff_addr;
+        int32_t buff_next = fw_enc_buff_next;
+
+        if ((fw_enc_buff_addr == 0) || (fw_enc_buff_addr == addr)) {
+            // Start of chunk been sent
+            buff_addr = addr;
+            fw_enc_buff_addr = addr;
+            memcpy(&fw_enc_buff, src8, len);
+            fw_enc_buff_next = addr + len;
+
+        } else
+
+        if (((addr - buff_addr) + len) > sizeof(fw_enc_buff)) {
+            // buffer's been filled, process existing buffer
+            // Copy new data into buffer afterwards
+            fw_enc_buff_addr = 0;
+
+        } else if (addr < buff_addr) {
+            // new addr chunk outside buffer, process existing buffer
+            // Copy new data into buffer afterwards
+            fw_enc_buff_addr = 0;
+
+        } else if (addr == buff_next) {
+            // Next normal chunk into buffer
+            memcpy(&fw_enc_buff.data[addr-buff_addr-sizeof(chunk_header_v1_t)], src8, len);
+            fw_enc_buff_next = addr + len;
+
+        } else {
+            // New addr range, process existing buffer
+            // Copy new data into buffer afterwards
+            fw_enc_buff_addr = 0;
+        }
+
+        size_t fw_enc_buff_len = fw_enc_buff_next - buff_addr;
+
+        if (fw_enc_buff.header.header_vers != 1) {
+            dfu_state.status = DFU_STATUS_ERROR_FILE;
+            dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+            return -1;
+        }
+        if (fw_enc_buff.header.address != buff_addr) {
+            // Chunk address doesn't agree with dfu address, abort
+            dfu_state.status = DFU_STATUS_ERROR_ADDRESS;
+            dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+            return -1;
+        }
+
+        if (fw_enc_buff.header.length > fw_enc_buff_len) {
+            // Don't have entire chunk yet
+            return 0;
+        }
+
+        if (hydro_sign_verify(fw_enc_buff.header.signature, fw_enc_buff.data, fw_enc_buff.header.length, ENC_CONTEXT, *mboot_public_key) != 0) {
+            // Signature failed
+            if (fw_enc_buff_addr == 0) { 
+                // We should have a full chunk here, if it fails we abort entire process
+                dfu_state.status = DFU_STATUS_ERROR_VERIFY;
+                dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+                return -1;
+            } else {
+                return 0;
```

Comment:
Is this a legitimate error condition? Should we assert?

Reply:
There were some normal use cases where it gets to here, I think it was related to not having the full amount of data yet (at start or end of flash/chunk) but I can't remember exactly what it was now. The not having a full chunk should in theory be picked up by the final if just above but I think there was something that it didn't catch.